### PR TITLE
Cache rewrite

### PR DESCRIPTION
**Pull request by ggaspersic**
Decided to re-write the cache functionality to make it more explicit and easier to follow what is happening & to prevent issues where Transformed features that were part of both the cached & non-cached part were being used as input.

Additionally also added reusable sub-methods in the FFM block that are always inlined & make the code dry-er.

This functionality was also already tested on a live system and didn't see any negative side effects both from performance and accuracy perspective.

### CODE REVIEW

File: src/regressor.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
 use std::any::Any;
+use std::collections::HashSet;
```
Comment (SkBlaz) [inline src/regressor.rs:3]: Same comment r.e. fx hash (rust compiler people really know their stuff)

**Review by SkBlaz — COMMENTED** [review]
Nice work!

File: src/block_ffm.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 use core::arch::x86_64::*;
 use std::any::Any;
+use std::collections::HashSet;
```
[the remark this answers is not in the document]
Comment (ggaspersic) [reply src/block_ffm.rs]: Nice one, totally forgot about that implementation. Replaced.

We should consider also doing that in some other places with the HashMap (separate PR 😄 )

File: src/block_ffm.rs [hunk 3]
```diff
@@ -709,7 +603,7 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
         unsafe {
             caches.push(BlockCache::FFM {
                 contra_fields: MaybeUninit::uninit().assume_init(),
-                contra_fields_present: vec![false; self.ffm_num_fields as usize],
+                features_present: HashSet::with_capacity(self.ffm_num_fields as usize),
```
[the remark this answers is not in the document]
Comment (ggaspersic) [reply src/block_ffm.rs]: At max all of them yes. Mind you this is called only once to allocate space. Afterwards its always cleared & the space is reused.

File: src/block_ffm.rs [hunk 4]
```diff
@@ -547,7 +493,7 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
             for field_index in 0..ffm_fields_count {
                 let field_index_ffmk = field_index * ffmk;
                 let field_index_ffmk_as_usize = field_index_ffmk as usize;
-                let offset = (field_index_ffmk * ffm_fields_count) as usize;
+                let offset = field_index_ffmk_as_usize * ffm_fields_count_as_usize;
```
[the remark this answers is not in the document]
Comment (ggaspersic) [reply src/block_ffm.rs:496]: One less conversion. We already do them so reusing the field makes sense.

File: src/regressor.rs (hunk 1 above)
Comment (ggaspersic) [reply src/regressor.rs]: Updated, tnx 👍

File: src/block_ffm.rs [hunk 5]
```diff
@@ -1029,7 +986,8 @@ mod tests {
             hash: 1,
             value: 1.0,
             contra_field_index: 0,
-        }]); // saying we have 1 field isn't entirely correct
+        }]);
+        // saying we have 1 field isn't entirely correct
```
[the remark this answers is not in the document]
Comment (ggaspersic) [reply src/block_ffm.rs:990]: Old comment will leave as is.

**Review by SkBlaz — APPROVED** [review]
Great, no further comments
